src/shims/fs.rs

```diff
@@ -582,6 +597,256 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
+
+        let result = read_dir(name);
+
+        match result {
+            Ok(dir_iter) => {
+                let size = 1;
+                let kind = MiriMemoryKind::Env;
+                let align = this.min_align(size, kind);
+                let dir_ptr = this.memory.allocate(Size::from_bytes(size), align, kind.into());
+                let prev = this
+                    .machine
+                    .dir_handler
+                    .streams
+                    .insert(dir_ptr, dir_iter);
+                if let Some(_) = prev {
+                    throw_unsup_format!("The pointer allocated for opendir was already registered by a previous call to opendir")
+                } else {
+                    Ok(Scalar::Ptr(dir_ptr))
+                }
+            }
+            Err(e) => {
+                this.set_last_error_from_io_error(e)?;
+                Ok(Scalar::from_int(0, this.memory.pointer_size()))
+            }
+        }
+    }
+
+    fn readdir64_r(
+        &mut self,
+        dirp_op: OpTy<'tcx, Tag>,
+        entry_op: OpTy<'tcx, Tag>,
+        result_op: OpTy<'tcx, Tag>,
+    ) -> InterpResult<'tcx, i32> {
+        let this = self.eval_context_mut();
+
+        this.check_no_isolation("readdir64_r")?;
+
+        let dirp = this.force_ptr(this.read_scalar(dirp_op)?.not_undef()?)?;
+
+        let entry_ptr = this.force_ptr(this.read_scalar(entry_op)?.not_undef()?)?;
+        let dirent64_layout = this.libc_ty_layout("dirent64")?;
+        this.memory.check_ptr_access(
+            Scalar::Ptr(entry_ptr),
+            dirent64_layout.size,
+            dirent64_layout.align.abi,
+        )?;
+
+        if let Some(dir_iter) = this.machine.dir_handler.streams.get_mut(&dirp) {
+            match dir_iter.next() {
+                Some(Ok(dir_entry)) => {
+                    // write into entry, write pointer to result, return 0 on success
+                    let entry_place = this.deref_operand(entry_op)?;
+                    let ino64_t_layout = this.libc_ty_layout("ino64_t")?;
+                    let off64_t_layout = this.libc_ty_layout("off64_t")?;
+                    let c_ushort_layout = this.libc_ty_layout("c_ushort")?;
+                    let c_uchar_layout = this.libc_ty_layout("c_uchar")?;
+
+                    let name_offset = dirent64_layout.details.fields.offset(4);
+                    let name_ptr = entry_ptr.offset(name_offset, this)?;
+
+                    #[cfg(unix)]
+                    let ino = std::os::unix::fs::DirEntryExt::ino(&dir_entry);
+                    #[cfg(not(unix))]
+                    let ino = 0;
+
+                    #[cfg(unix)]
+                    let file_name = dir_entry.file_name();
+                    #[cfg(unix)]
+                    let file_name = std::os::unix::ffi::OsStrExt::as_bytes(file_name.as_os_str());
+                    #[cfg(not(unix))]
+                    let file_name = b"";
```

Comment:
Uh, on non-Unix we don't even have a *file name*? Doesn't that make the entire thing useless?

Reply:
Ah I think this entire `cfg`-then-else here should be gone, you shouldn't do any OsStr handling here. Leave that to the helpers. `write_os_str_to_c_str` should work on `file_name.as_os_str()`.